nixos/modules/services/networking/unifi.nix

```diff
@@ -140,50 +198,70 @@ in
     systemd.mounts = map ({ what, where }: {
         bindsTo = [ "unifi.service" ];
         partOf = [ "unifi.service" ];
-        unitConfig.RequiresMountsFor = stateDir;
         options = "bind";
-        what = what;
-        where = where;
+        inherit what where;
       }) mountPoints;
 
-    systemd.tmpfiles.rules = [
-      "e '${stateDir}' 0700 unifi - - -"
-      "d '${stateDir}/data' 0700 unifi - - -"
-    ];
+    systemd.services.unifi-setup = {
+      description = "UniFi controller daemon - setup";
+      wantedBy = [ "multi-user.target" ];
+      before = [ "unifi.service" ];
+
+      script = let
```

Comment:
We cannot drop the tmpfiles rules. The main unit pulls in the bindmounts which are mounted under /var/lib/unifi, so that directory needs to exist before. I lost about an hour of my life because I couldn't understand why it refused to start with statedirectory.... That actually leads me to another point. Any good ideas on how to document the implementation details? Just comments inside?

As for the ```unifi-setup``` vs using ```preStart```, it's become a pattern I have started using (I'll do an RFC on it) because it gives more flexibility. Once preStart has been set, you cannot add to it from the outside if the service needs to be augmented whereas using a separate -init unit does. Additionally, we should then be pulling in a ```unifi.target```. Anyway, I'll do a separate write-up on my thoughts on this.

Reply:
That makes sense. Thanks for the explanation.

I look forward to reading such an RFC.